modules/webgl/src/adapter/resources/webgl-command-buffer.ts

```diff
@@ -235,93 +242,72 @@ function _copyTextureToTexture(device: WebGLDevice, options: CopyTextureToTextur
 
   let {
     width = options.destination.width,
-    height = options.destination.width
+    height = options.destination.height
     // depthOrArrayLayers = 0
   } = options;
 
-  const destinationMipmaplevel = 0;
-  const destinationInternalFormat = GL.RGBA;
-
   const {framebuffer, destroyFramebuffer} = getFramebuffer(source);
   const [sourceX, sourceY] = origin;
+  const [destinationX, destinationY, destinationZ] = destinationOrigin;
 
-  const isSubCopy = false;
-  // typeof destinationX !== 'undefined' ||
-  // typeof destinationY !== 'undefined' ||
-  // typeof destinationZ !== 'undefined';
-
-  // destinationX = destinationX || 0;
-  // destinationY = destinationY || 0;
-  // destinationZ = destinationZ || 0;
-  device.gl.bindFramebuffer(GL.FRAMEBUFFER, framebuffer.handle);
+  // @ts-expect-error native bindFramebuffer is overridden by our state tracker
+  const prevHandle: WebGLFramebuffer | null = device.gl.bindFramebuffer(
+    GL.FRAMEBUFFER,
+    framebuffer.handle
+  );
   // TODO - support gl.readBuffer (WebGL2 only)
   // const prevBuffer = gl.readBuffer(attachment);
 
-  let texture = null;
+  let texture: WEBGLTexture = null;
   let textureTarget: GL;
   if (destination instanceof WEBGLTexture) {
     texture = destination;
     width = Number.isFinite(width) ? width : texture.width;
     height = Number.isFinite(height) ? height : texture.height;
     texture.bind(0);
-    textureTarget = texture.destination;
+    textureTarget = texture.glTarget;
   } else {
-    throw new Error('whoops');
-    //  textureTarget = destination;
+    throw new Error('invalid destination');
   }
 
-  if (!isSubCopy) {
-    device.gl.copyTexImage2D(
-      textureTarget,
-      destinationMipmaplevel,
-      destinationInternalFormat,
-      sourceX,
-      sourceY,
-      width,
-      height,
-      0 /* border must be 0 */
-    );
-  } else {
-    // switch (textureTarget) {
-    //   case GL.TEXTURE_2D:
-    //   case GL.TEXTURE_CUBE_MAP:
-    //     device.gl.copyTexSubImage2D(
-    //       textureTarget,
-    //       destinationMipmaplevel,
-    //       destinationX,
-    //       destinationY,
-    //       sourceX,
-    //       sourceY,
-    //       width,
-    //       height
-    //     );
-    //     break;
-    //   case GL.TEXTURE_2D_ARRAY:
-    //   case GL.TEXTURE_3D:
-    //     device.gl.copyTexSubImage3D(
-    //       textureTarget,
-    //       destinationMipmaplevel,
-    //       destinationX,
-    //       destinationY,
-    //       destinationZ,
-    //       sourceX,
-    //       sourceY,
-    //       width,
-    //       height
-    //     );
-    //     break;
-    //   default:
-    // }
+  switch (textureTarget) {
+    case GL.TEXTURE_2D:
+    case GL.TEXTURE_CUBE_MAP:
+      device.gl.copyTexSubImage2D(
```

Comment:
There are new helper functions for 9.1 in webgl-texture-helpers that take care of calling the right webgl function and handle more cases, but if this is meant for cherry-pick to 9.0 the maybe aligning with those can be a separate step.